pkg/pillar/cmd/nodeagent/serverdns.go

```diff
@@ -0,0 +1,133 @@
+	if len(etchosts) == 0 || len(cloudServer) == 0 {
+		return ipaddrCached, maxTTLSec
+	}
+
+	// Check to see if the server domain is already in the /etc/hosts as in eden,
+	// then skip this DNS queries
+	if ipaddrCached == "" {
+		hostsEntries := bytes.Split(etchosts, []byte("\n"))
+		for _, entry := range hostsEntries {
+			fields := bytes.Fields(entry)
+			if len(fields) == 2 {
+				if bytes.Compare(fields[1], cloudServer) == 0 {
+					log.Tracef("server entry %s already in /etc/hosts, skip", cloudServer)
+					return ipaddrCached, maxTTLSec
+				}
+			}
+		}
+	}
+
+	var nameServers []string
+	dnsServer, _ := ioutil.ReadFile(resolvFileName)
+	dnsRes := bytes.Split(dnsServer, []byte("\n"))
+	for _, d := range dnsRes {
+		d1 := bytes.Split(d, []byte("nameserver "))
+		if len(d1) == 2 {
+			nameServers = append(nameServers, string(d1[1]))
+		}
+	}
+	if len(nameServers) == 0 {
+		nameServers = append(nameServers, "8.8.8.8")
+	}
+
+	if _, err := os.Stat(tmpHostFileName); err == nil {
+		os.Remove(tmpHostFileName)
+	}
+
+	var newhosts []byte
+	var gotipentry bool
+	var lookupIPaddr string
+	var ttlSec int
+
+	domains := []string{string(cloudServer)}
+	dtypes := []dns.QueryType{dns.TypeA}
+	for _, nameServer := range nameServers {
+		resolver := dns.NewResolver(nameServer)
+		resolver.Targets(domains...).Types(dtypes...)
+
+		res := resolver.Lookup()
+		for target := range res.ResMap {
+			for _, r := range res.ResMap[target] {
+				dIP := net.ParseIP(r.Content)
+				if dIP == nil {
+					continue
+				}
+				lookupIPaddr = dIP.String()
+				ttlSec = getTTL(r.Ttl)
+				if ipaddrCached == lookupIPaddr {
+					log.Tracef("same IP address %s, return", lookupIPaddr)
+					return ipaddrCached, ttlSec
+				}
+				serverEntry := fmt.Sprintf("%s %s\n", lookupIPaddr, cloudServer)
+				newhosts = append(etchosts, []byte(serverEntry)...)
+				gotipentry = true
+				// a rare event for dns address change, log it
+				log.Noticef("dnsServer %s, ttl %d, entry add to /etc/hosts: %s", nameServer, ttlSec, serverEntry)
+				break
+			}
+			if gotipentry {
+				break
+			}
+		}
+		if gotipentry {
+			break
+		}
+	}
+
+	if ipaddrCached == lookupIPaddr {
+		return ipaddrCached, minTTLSec
+	}
+	if !gotipentry { // put original /etc/hosts file back
+		newhosts = append(newhosts, etchosts...)
+	}
+
+	ipaddrCached = ""
+	err := ioutil.WriteFile(tmpHostFileName, newhosts, 0644)
+	if err == nil {
+		if err := os.Rename(tmpHostFileName, etcHostFileName); err != nil {
+			log.Errorf("can not rename /etc/hosts file %v", err)
+		} else {
+			if gotipentry {
+				ipaddrCached = lookupIPaddr
+			}
+			log.Tracef("append controller IP %s to /etc/hosts", lookupIPaddr)
+		}
+		os.Remove(tmpHostFileName)
+	} else {
+		log.Errorf("can not write /tmp/etchosts file %v", err)
+	}
+
+	return ipaddrCached, ttlSec
+}
+
+func getTTL(ttl time.Duration) int {
+	ttlSec := int(ttl.Seconds())
+	if ttlSec < minTTLSec {
+		ttlSec = minTTLSec
```

Comment:
Probably makes sense to have a log.Functionf when we clamp the ttl to the min; log the name, IP. and received ttl.

Reply:
Ok.

Reply:
actually i found previous i was using a google wifi router, which returns the ttl of the query as 300 reliably, but the one in the lab, actually returns the remaining value of it's cached ttl time, which we often hit the very small value. So it seems to be a normal case in certain dns server implementation.